## **Numpy For Data Scince**

Importing Numpy


In [ ]:
import numpy as np

**Numpy Vs Python**

1. Speed and Performance:
NumPy: Much faster when working with large arrays and numerical computations due to its implementation in C and Fortran.
Python: Native Python lists and loops are slower because they are dynamically typed and rely on interpreted execution.
2. Memory Efficiency:
NumPy: Uses less memory because it stores data in contiguous memory blocks, and data types are fixed, unlike Python’s general-purpose lists.
Python: Lists store objects and pointers, leading to more memory overhead and slower access times.
3. Multidimensional Arrays:
NumPy: Provides the ndarray, which can efficiently handle multidimensional data (2D, 3D, or more). This is a cornerstone of numerical and scientific applications.
Python: You’d have to use nested lists to achieve similar functionality, but it's not efficient and lacks built-in functions for advanced operations.
4. Mathematical Operations:
NumPy: Provides a wide range of optimized mathematical functions such as linear algebra, Fourier transformations, and random number generation. Operations are element-wise, vectorized, and optimized.
Python: Native lists require explicit loops for element-wise operations, which are slower and harder to write.**bold text**

In [ ]:
#using pyhton for lists
import time
a = [123,234,123]
b = [4, 5, 6]
starting_time = time.time()
result = [a[i] + b[i] for i in range(len(a))]
ending_time = time.time()
sum_Of_time_list = (ending_time - starting_time)*100
print("Sum of lists : ",result)
print(f"the time for list is : {sum_Of_time_list:.2f} Milliseconds")

#using numpy for arrays
import numpy as np
a = np.array([1, 2, 3])
b = np.array([4, 5, 6])
starting_time = time.time()
result = a + b  # Vectorized operation
ending_time = time.time()
sum_Of_time_arrey = (ending_time - starting_time)*100
print(result)
print(f"the time for arrey is : {sum_Of_time_arrey:.2f} Milliseconds")



Sum of lists :  [127, 239, 129]
the time for list is : 0.01 Milliseconds
[5 7 9]
the time for arrey is : 0.01 Milliseconds


# **Numpy Arrays**

Key Features of a NumPy Array:
Fixed Size: Once created, the size of a NumPy array cannot be changed.
Homogeneous: All elements in a NumPy array must be of the same data type (e.g., all integers or all floats).
Fast and Memory-Efficient: NumPy arrays consume less memory and provide faster access compared to Python lists.
Multi-Dimensional: NumPy arrays can be 1-dimensional (vectors), 2-dimensional (matrices), or n-dimensional (nD-arrays). *italicized text*

In [ ]:
array1 = np.arange(1,10)#staring,ending and stepping
i = array1
#for array1 in i > 5:
    #print(i*0)
print(array1)



[1 2 3 4 5 6 7 8 9]


NumPy arrays are much faster, more memory-efficient, and provide rich functionality for mathematical operations.
They allow for efficient manipulation of large datasets, broadcasting, and multidimensional data handling.

**One Dimensional Array**

In [ ]:
#one dimensional arrays using numpy
arr1 = np.array([1,4,61,222])
#for i in arr1:

print(arr1)


[  1   4  61 222]


**Two Dimensional Array**

In [ ]:
#Two dimensional array using numpy
arr2 = np.array([[1,77,33],[6,33,11]])
print(arr2)
#time_taken = time.time()

#print(time_taken , " in milliseconds")

[[ 1 77 33]
 [ 6 33 11]]


**Multi Dimensional Array using Numpy**

In [ ]:
arr3 = np.array([[1,77,33],[6,33,11],[3,77,55],])
print(arr3)

[[ 1 77 33]
 [ 6 33 11]
 [ 3 77 55]]


# **Properties of Array using numpy**

NumPy arrays have the following key properties:

ndim: The number of dimensions (axes) of the array.

shape: The shape of the array, represented as a tuple of integers indicating the size of the array in each dimension.

size: The total number of elements in the array (product of its dimensions).

dtype: The data type of the elements in the array, such as int32, float64, etc.

itemsize: The size of each element in the array (in bytes).

nbytes: The total memory consumed by the array (in bytes), equal to the product of size and itemsize.

T: The transpose of the array (rows become columns).

flat: A flat iterator over all elements in the array, treating it as if it were 1-dimensional.

real and imag: The real and imaginary parts of the array if it contains complex numbers.

In [ ]:
print("The number of dimensions (axes) of the array : ",arr3.ndim)
print("shape : ",arr3.shape)
print("size of an array is : ",arr3.size)
print("dataType : ",arr3.dtype)
print("Itemsize of an array : ",arr3.itemsize)
print("The total memory consumed by the array : ",arr3.nbytes)
print("The transpose of the array : ",arr3.T)
print("Flate the array : " ,arr3.flat)

The number of dimensions (axes) of the array :  2
shape :  (3, 3)
size of an array is :  9
dataType :  int64
Itemsize of an array :  8
The total memory consumed by the array :  72
The transpose of the array :  [[ 1  6  3]
 [77 33 77]
 [33 11 55]]
Flate the array :  <numpy.flatiter object at 0x567d2dffb2f0>


In [ ]:
#3D array with Zeros
Zeroth_Array = np.zeros((2,5,5))
print(Zeroth_Array)

[[[0. 0. 0. 0. 0.]
  [0. 0. 0. 0. 0.]
  [0. 0. 0. 0. 0.]
  [0. 0. 0. 0. 0.]
  [0. 0. 0. 0. 0.]]

 [[0. 0. 0. 0. 0.]
  [0. 0. 0. 0. 0.]
  [0. 0. 0. 0. 0.]
  [0. 0. 0. 0. 0.]
  [0. 0. 0. 0. 0.]]]


In [ ]:
Once_Array = np.ones((2,5,5))
print(Once_Array)

[[[1. 1. 1. 1. 1.]
  [1. 1. 1. 1. 1.]
  [1. 1. 1. 1. 1.]
  [1. 1. 1. 1. 1.]
  [1. 1. 1. 1. 1.]]

 [[1. 1. 1. 1. 1.]
  [1. 1. 1. 1. 1.]
  [1. 1. 1. 1. 1.]
  [1. 1. 1. 1. 1.]
  [1. 1. 1. 1. 1.]]]


# **DataTypes in Numpy**

1. Integer Types:
int8: Integer (-128 to 127), 1 byte
int16: Integer (-32,768 to 32,767), 2 bytes
int32: Integer (-2,147,483,648 to 2,147,483,647), 4 bytes
int64: Integer (-9,223,372,036,854,775,808 to 9,223,372,036,854,775,807), 8 bytes
2. Unsigned Integer Types:
uint8: Unsigned integer (0 to 255), 1 byte
uint16: Unsigned integer (0 to 65,535), 2 bytes
uint32: Unsigned integer (0 to 4,294,967,295), 4 bytes
uint64: Unsigned integer (0 to 18,446,744,073,709,551,615), 8 bytes
3. Floating-Point Types:
float16: Half precision float, 2 bytes
float32: Single precision float, 4 bytes
float64: Double precision float, 8 bytes
4. Complex Number Types:
complex64: Complex number, represented by two 32-bit floats (real and imaginary parts), 8 bytes
complex128: Complex number, represented by two 64-bit floats (real and imaginary parts), 16 bytes
5. Boolean Type:
bool: Boolean (True or False), 1 byte
6. String Types:
string_: Fixed-size string data. Length determined by the number following the S, e.g., S10 for 10 characters.
unicode_: Fixed-size Unicode data. Length determined by the number following the U, e.g., U10 for 10 characters.

In [ ]:
#int datatypes
array_int = np.array([1,5,56,2],dtype=np.int32)
int_array = np.array([1,5,56,2],dtype=np.int64)
print(int_array.dtype)
print(array_int.dtype)
array_bool = np.array([1,5,56,2],dtype=np.bool_)
print(array_bool.dtype)
array_float = np.array([1,5,56,2],dtype=np.float64)
print(array_float.dtype)

int64
int32
bool
float64


# **Operators in Numpy**

**Arithmetic Operations:**


These are performed element-wise and are vectorized for performance:

Addition: +
Subtraction: -
Multiplication: *
Division: /
Exponentiation: **
Modulus:

In [ ]:
arr_addition = np.add([1,55,22,56],[4,77,2,95])
print("addition of array ",arr_addition)
arr_subtraction = np.subtract([1,55,22,56],[4,77,2,95])
print("subtraction of array ",arr_subtraction)
arr_multiplication = np.multiply([1,55,22,56],[4,77,2,95])
print("multiplication of array ",arr_multiplication)
arr_division = np.divide([1,55,22,56],[4,77,2,95])
print("Division of array ",arr_division)
arr_exponentiation = np.power([1,55,22,56],[4,77,2,95])
print("Exponatial of array ",arr_exponentiation)
arr_modulus = np.mod([1,55,22,56],[4,77,2,95])
print("Modulus of array ",arr_modulus)



addition of array  [  5 132  24 151]
subtraction of array  [ -3 -22  20 -39]
multiplication of array  [   4 4235   44 5320]
Division of array  [ 0.25        0.71428571 11.          0.58947368]
Exponatial of array  [                   1 -7557483726025653929                  484
                    0]
Modulus of array  [ 1 55  0 56]


**Mathematical Operators in Numpy**

Mathematical Functions:
NumPy offers a wide range of mathematical functions that operate element-wise:

np.sqrt(): Square root
np.exp(): Exponential (e^x)
np.log(): Natural logarithm
np.sin(), np.cos(), np.tan(): Trigonometric functions
np.sum(): Sum of array elements
np.mean(): Mean of array elements
np.median(): Median of array elements
np.std(): Standard deviation of array elements
np.max(), np.min(): Maximum and minimum values in an array

In [ ]:
array_mean = np.mean([1,55,22,56])
print("Mean of array ",array_mean)
array_median = np.median([1,55,22,56])
print("Median of array ",array_median)
array_standard_deviation = np.std([1,55,22,56])
print("Standard Deviation of array ",array_standard_deviation)
array_cos = np.cos([1,55,22,56])
print("Cos of array ",array_cos)
array_sin = np.sin([1,55,22,56])
print("Sin of array ",array_sin)
array_tan = np.tan([1,55,22,56])
print("Tan of array ",array_tan)
array_sum = np.sum([1,55,22,56])
print("Sum of array ",array_sum)
array_max = np.max([1,55,22,56])
print("Max of array ",array_max)
array_min = np.min([1,55,22,56])
print("Min of array ",array_min)


Mean of array  33.5
Median of array  38.5
Standard Deviation of array  23.221757039466244
Cos of array  [ 0.54030231  0.02212676 -0.99996083  0.85322011]
Sin of array  [ 0.84147098 -0.99975517 -0.00885131 -0.521551  ]
Tan of array  [ 1.55740772e+00 -4.51830879e+01  8.85165604e-03 -6.11273688e-01]
Sum of array  134
Max of array  56
Min of array  1


# **Indexing and Slicing**

**indexing and Slicing in 1D array**



In [33]:
import numpy as np
array_indexing = np.array([1,55,22,56])
#print(array_indexing[1:3:2])
array_indexing[1] = 100
#print(array_indexing)
array_indexing[2] = 200
print(array_indexing)
#slicing
arr = np.array([10, 20, 30, 40, 50])
print(arr[1:4])  # Output: [20 30 40] (elements from index 1 to 3)
print(arr[::2])  # Output: [10 30 50] (every second element)


[  1 100 200  56]
[20 30 40]
[10 30 50]


**indexing and slicing in 2D array**

In [34]:
arr = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
print(arr[2, 2])  # Output: 6 (element in the 2nd row and 3rd column)
#slicing in 2D array
print(arr[0:2,1:3])
print(arr[0:3,1:2])


9
[[2 3]
 [5 6]]
[[2]
 [5]
 [8]]


**indexing and slicing in 3D array**

In [32]:
arr = np.array([[[1, 2],
                 [3, 4]],
                [[5, 6],
                 [7, 8]]])
print(arr[0, 1, 1])  # Output: 4 (element in the 1st block, 2nd row, 2nd column)


4


**Boolean indexing**

In [40]:
arr = np.array([10, 20, 30, 40, 50])
print(arr[arr >= 40])  # Output: [40 50]


[40 50]


In [41]:
arr = np.random.normal(0, 1, 5)
print(arr)


[0.39574707 0.6083932  0.74163259 0.00220765 0.68293463]


In [42]:
arr = np.random.normal(10, 2, (2, 3))
print(arr)


[[ 8.56150317 10.83395752  9.07833359]
 [ 5.5561304  11.38778105 10.58464736]]


# **Masking**

In [43]:
import numpy as np

arr = np.array([10, 20, 30, 40, 50])
mask = arr > 30  # Create a mask where elements are greater than 30
print(mask)  # Output: [False False False  True  True]


[False False False  True  True]


In [44]:
arr = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
mask = arr > 5  # Create a mask for elements greater than 5
print(mask)
filtered_arr = arr[mask]
print(filtered_arr)  # Output: [6 7 8 9]


[[False False False]
 [False False  True]
 [ True  True  True]]
[6 7 8 9]


In [46]:
arr = np.array([10, 20, 30, 40, 50])
mask = (arr > 20) & (arr < 50)  # Select elements greater than 20 and less than 50
print(mask)  # Output: [False  True  True  True False]
filtered_arr = arr[mask]
print(filtered_arr)  # Output: [30 40]


[False False  True  True False]
[30 40]


In [47]:
arr = np.array([10, 20, 30, 40, 50])
result = np.where(arr > 30, arr, -1)  # Replace elements <= 30 with -1
print(result)  # Output: [-1 -1 -1 40 50]


[-1 -1 -1 40 50]


# **Ststistical Functions**

In [49]:
array = np.array([1,55,22,56,66,11,45])
print(np.mean(array))
print(np.median(array))
print(np.std(array))
print(np.var(array))
print(np.min(array))
print(np.max(array))

36.57142857142857
45.0
23.255457994428152
540.8163265306122
1
66


#***spliting and joining***

In [57]:
array = np.arange(1,16)
array1 = np.split(array,3)
print(array1)

[array([1, 2, 3, 4, 5]), array([ 6,  7,  8,  9, 10]), array([11, 12, 13, 14, 15])]


**Horizontal split and vertical splitting**

In [63]:
arr = np.array([[1, 2, 3,2],
                [4, 5, 6,5],
                [7, 8, 9,7]])
horizontal_split = np.hsplit(arr, 4)
print(horizontal_split)
vertical = np.vsplit(arr,3)
print(vertical)

[array([[1],
       [4],
       [7]]), array([[2],
       [5],
       [8]]), array([[3],
       [6],
       [9]]), array([[2],
       [5],
       [7]])]
[array([[1, 2, 3, 2]]), array([[4, 5, 6, 5]]), array([[7, 8, 9, 7]])]


**concatination**

In [67]:
arr_1 = np.arange(1,50)
arr_2 = np.array([4, 5, 6])
concatenated_array = np.concatenate((arr_1, arr_2))
print(concatenated_array)

[ 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24
 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48
 49  4  5  6]


**adding and deleting using Numpy**

In [75]:
arr_2 = np.array([4, 5, 6])
appended = np.append(arr_2,1)
appended_array = np.append(arr_2, [7, 8, 9])
print(appended_array)
deleted = np.delete(appended,1)
print(deleted)
#in 2D array
arr = np.array([[1, 2, 3],
                [4, 5, 6]])
appended_array = np.append(arr, [[7, 8, 9]], axis=0)
print(appended_array)
deleted = np.delete(arr, 1, axis=0)
print(deleted)

[4 5 6 7 8 9]
[4 6 1]
[[1 2 3]
 [4 5 6]
 [7 8 9]]
[[1 2 3]]
